### PR TITLE
wip: add Rubocop and Linting lesson

### LINKED ISSUE #27091: Ruby Course: Add Rubocop / Linting Lesson
### Discussed in https://github.com/TheOdinProject/curriculum/discussions/25674

<div type='discussions-op-text'>

<sup>Originally posted by **ChargrilledChook** March 14, 2022</sup>

Complete the following REQUIRED checkboxes:
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this issue follows the `location for request: brief description of request` format, e.g. `NodeJS course: Add lessons on XYZ`

The following checkbox is OPTIONAL:

-   [x] I would like to be assigned this issue to work on it

<hr>

**1. Description of the Feature Request:**

Rubocop is a very useful tool that is regularly recommended in Discord (as well as briefly introduced in the OOP section). 

However, it is often also a source of confusion and questions in Discord, with many learners not understanding the difference between Rubocop the gem and the IDE integration. We also give very little context on how to read the output.

Introducing a lesson that covered topics such as installation and configuration, how to interpret / look up offenses on the community style guide and basic customization could provide a lot of value to learners, as well as provide parity with the linting lessons in the JS section.

**2. Acceptance Criteria:**
A lesson that covers:
-  [ ] Gem installation + IDE integration
-  [ ] Difference between CLI and IDE versions, visual examples
-  [ ] Brief explanation of linting
-  [ ] Intro to community style guide + difference between different types of rules (style, lint, performance etc)
-  [ ] Explanation of more confusing rules (such as ABC / cyclomatic complexity metrics)
-  [ ] How to write a custom .rubocop.yml / add plugins

**3. Additional Information:**

If the maintainers like this idea I'm happy to create a more detailed lesson outline for discussion  and/or provide a draft.
</div>

### PR DESCRIPTION
**Pull request by scheals**
## Because
We never had a dedicated lesson for Linting and Rubocop. This is even more important now because of Ruby LSP requiring a locally installed RuboCop for linting and formatting to work.

## This PR
- Adds RuboCop and Linting lesson.
- Removes outdated RuboCop installation instructions in Object Oriented Programming lesson.
- Does some style cleanup in the aforementioned OOP lesson.

## Issue

Closes #27091

## Pull Request Requirements
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [ ] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by scheals** [comment]
@JoshDevHub 
First draft is ready. Assignment section and Additional resources is very WIP but the foundations are there.

Be frank with me and I hope that the content looks good :D

**Comment by JoshDevHub** [comment]
I'm on triage this week, so it might take me a few days to read through this more thoroughly. Looking pretty good though! Will be awesome to have this content in the curriculum.

**Review by JoshDevHub — COMMENTED** [review]
Just a couple of comments so far. I think it's pretty awesome. Let me know if you want any input on any of the Todo items.

File: ruby/object_oriented_programming_basics/linting_and_rubocop.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+       ^^^
+
+2 files inspected, 16 offenses detected, 14 offenses corrected, 2 more offenses can be corrected with rubocop -A
+# Duplicate offenses were truncated.
+```
+
+See that `.`? That means the first file is now all fine and dandy! Some of the offenses were not corrected by RuboCop and that's because `-a` is for safe autocorrect.
+If you wanted to go through with the `[Correctable]` offenses, you'd want to use `-A` as the output helpfully suggests. This is due to the fact that some Cops are safe, some are unsafe.
+
+The safe Cops promise that they won't have false positives and that their autocorrect is won't change the semantics of the code and it will be fully equivalent to what you've had written.
```
[the remark this answers is not in the document]
Comment (scheals) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md]: Thanks for catching this :)

File: ruby/object_oriented_programming_basics/linting_and_rubocop.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+From this it follows that unsafe Cops either have false positives or slightly change the semantics of the code. The first characteristic means they tell you something is wrong when actually, everything is alright - for example, you've got a method with the same name as one in standard library in your object and RuboCop treats it as if it were the standard library method.
+
+<!--
+TODO: find out whether https://docs.rubocop.org/rubocop/usage/autocorrect.html talks about human semantics or the machine semantics.
+The second example seems to suggest the latter - autocorrecting breaks the code.
```
[the remark this answers is not in the document]
Comment (scheals) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md:173]: Neat, cheers for clarification. :) I somehow never registered the explanations for safety in Cop documentation. Useful to know!

File: ruby/object_oriented_programming_basics/linting_and_rubocop.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  NewCops: enable
+```
+
+in your `.rubocop.yml` to enable all the new Cops.
+
+### Power corrupts
+
+Perhaps you're not interested in tailoring RuboCop to your liking, especially since you're just starting out and have absolutely no idea what's good and what's bad. That'd be the correct approach - don't worry about it right now and just go with the RuboCop defaults.
+
+There exists another set of rules which is executed by a wrapper over RuboCop called Standard Ruby. Standard aims at providing one opinionated set of rules, smaller in scope than RuboCop, in hopes that it will provide a standard for all Rubyists to follow.
```
[the remark this answers is not in the document]
Comment (scheals) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md:206]: I'm not sure either. My rationale for the inclusion was mostly the fact that I could foresee people going into some deep rabbit hole of RuboCop configuration or grabbing config off some company just for the sake of having a config that is not default. This might have some bad consequences like in my mind like the lack of Metrics cops mentioned in the draft.

Buuut just mentioning Standard might do just that, even though I've made a (hopefully) strong case for sticking with RuboCop. I'm OK with dropping mention of Standard completely, so I wouldn't even put it in additional resources.

**Comment by scheals** [comment]
> Just a couple of comments so far. I think it's pretty awesome. Let me know if you want any input on any of the Todo items.

Thanks!

Hmmm, I haven't really directly defined linting and formatting and I wonder if that's a problem. Formatting is probably clear enough on its own but perhaps it'd be good to make a clear distinction between the two? Perhaps just linking to the wiki page of Linting in the assignment is going to be enough. Not sure.

Also not super confident in my ABC/Cyclomatic/Perceived metric explanations.

(Hope this is on-topic, this is how I understood Todo items :) )

**Comment by scheals** [comment]
Interestingly enough, I had some typos/misspellings and while the action has pointed out them in the preview, the action in the summary has not failed. Not sure if intentional.

image

**Comment by nguyenlekhtn** [comment]
HI, "This is even more important now because of Ruby LSP requiring a locally installed RuboCop for linting and formatting to work."

As I checked the latest Rbuy-LSP version, it is no longer true. It doesn't require rubucop when I create new project.

**Comment by scheals** [comment]
@nguyenlekhtn 

That sounds great! Could you provide me with an example file that I can test this with? My tests went the usual route with Ruby-LSP:
1. It fails to do anything without a Gemfile.
2. Closing the tab with a Ruby file and reopening it gets it to ask me about monorepo setup.
3. It starts finally.
4. I write some code that RuboCop should hate - it is silent.

Perhaps you need to have a default linter/formatter set up? I have none and rely on the Gemfile:
image

For more context, I cloned TOP's ruby-exercises repo, opened it in VSCode and then started playing around:
image
image

**Comment by nguyenlekhtn** [comment]
Hi @scheals, apologize, look like I misread. What I mean is now Ruby-LSP run without a Gemfile. It still requires rubocop to lint and format, as expected.

**Comment by scheals** [comment]
Ahh no worries. Yeah, in the meantime Ruby LSP fixed it not working without a Gemfile but I still find its activation rather hit and miss - often I open a Ruby project and Ruby LSP only opens after I close-reopen a Ruby file for it to pick up that I'm working on a Ruby file. I'm expecting looots of questions about this - I imagine that the bit I have proposed to change in `Installing Ruby`:
> The most important features Ruby LSP provides will work out of the box. But it may bug you about using a monorepo setup, missing lockfiles or rubocop - you can choose "Don't show again" for now. We will introduce these later.

Will not be sufficient and instructions how to make Ruby LSP start without a Gemfile are going to be needed.

File: ruby/object_oriented_programming_basics/linting_and_rubocop.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoshDevHub) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md:202]: I think the config rabbit hole can be avoided by doing two things:
1. Mentioning that people should stick to rubocop's defaults when starting out. You kind of have this going on already with the paragraph starting on lines 204, but I think it's good advice for people who are starting out
2. I think it'd also be potentially good to show how to configure the two most common things people have problems with on rubocop's default settings: the magic comment for frozen string literals and using double instead of single quotes for strings.

What do you think?

[the remark this answers is not in the document]
Comment (scheals) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md:202]: So hammer the point mentioned in 1. harder + show how to throw magic comment/quote string rules into `rubocop.yml`?

**Comment by scheals** [comment]
So, in the `Lesson overview` this promise is made:
> - What the differences between RuboCop in the CLI and in your VSCode are.

and I am not sure whether I've kept it as of current draft. The lesson does introduce how to use RuboCop in VSCode but doesn't really talk about the differences in what RuboCop in the CLI can do and what RuboCop integration in VSCode can't do and vice versa. Perhaps this should be changed to:
> - How to use RuboCop in VSCode.

which would be OK for the current content.

I have "finalized" the assignment and the additional resources and have proposed a rather long list of knowledge checks. I took a quick glance at the freshly revamped React course and I found at least one lesson with similar check count, so perhaps it is A-OK and I can start filling in the anchors.

Strongly believe that including Sandi Metz' talk on Rules in the Assignment section is the way to go. I gave it a re-listen and it is even better than I remembered it. Thrown out a link to RuboCop's GH repo as the docs would have the same information anyway.

Material related to Standard was also removed although I'm still waiting for clarification as to what is needed to be done with the mention of Standard in the lesson content itself - I'll clear that up together with the common config issues then.

File: ruby/object_oriented_programming_basics/linting_and_rubocop.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoshDevHub) [reply ruby/object_oriented_programming_basics/linting_and_rubocop.md:202]: Yep!

I think this would help people avoid too much config rabbit holing. We can potentially revisit though if config bikeshedding seems to be a problem in discord conversations.

**Comment by JoshDevHub** [comment]
This lesson looks great @scheals 

I'm a fan of your knowledge checks, so feel free to go ahead and start filling in the anchors whenever you want. Also love that Sandi Metz talk and I'm glad to have it in the assignment.

Was there any other specific thing you were concerned about or want me to respond to?

**Comment by scheals** [comment]
I'm very glad to hear that @JoshDevHub :D 

So, I had to include an explicit definition of linting and formatting in the lesson itself, otherwise the anchors would not be as straightforward and might've lead to confusion. I've dropped mention of Standard and brought instructions on how to change particular rules and how to inherit configuration from other `.rubocop.yml` files. I am not sure whether they should live where they do or whether they should come after/before `.rubocop.yml` explanation and enabling NewCops.

I wonder if the install/run knowledge checks aren't too close to each other and overall too laconic:
>  Go back and install RuboCop locally (as in, use Bundler) and then run bundle exec rubocop in your terminal.

On one hand, it is just that easy and this knowledge should be present due to the previous Bundler lesson but I wonder if those two knowledge checks should be merged into one. But perhaps the questions themselves being there are important, to check whether folks actually know.

As for requests - please check the Metrics section to make sure it is clear and correct :)

In short:
- Is the order for `.rubocop.yml` configuration and inline configuration OK? (general overview of the config file, then inline, then more detailed explanation of configuring)?
- Is the Metrics section good?
- Do knowledge checks make sense where they are and is content they link to enough?

**Comment by JoshDevHub** [comment]
> Is the order for .rubocop.yml configuration and inline configuration OK? (general overview of the config file, then inline, then more detailed explanation of configuring)?

I think the order is great. The lesson flows very well in my opinion.

There may be an argument later to split this up (the same way we're thinking of splitting up the bundler / project organization lesson). Maybe there can be a Introduction/setup lesson and then a lesson for configuration? I'm fine to shift this to a different PR though.

> Is the Metrics section good?

I think it's great! Wish I had it while I was learning :laughing: 

> Do knowledge checks make sense where they are and is content they link to enough?

They all seem good to me :thumbsup:

**Comment by JoshDevHub** [comment]
@scheals Can you try to merge upstream `main` and see if that fixes whatever is happening with the linter action?

Also feel free to mark this as ready for review.

One thing was curious for your thoughts on: where do you think the two new lessons should be positioned within the Ruby course?

**Comment by scheals** [comment]
Linter is not happy but I don't think it needs to be corrected:
image
This is about: 
image

As for placement, I think they should be introduced right after Object Oriented Programming, of course Bundler lesson before RuboCop one. In the future I could see that project that was thought about way in the past that'd be about extending existing app would come before the RuboCop lesson. If that ever comes to fruition, of course.

**Review by JoshDevHub — APPROVED** [review]
I agree with that placement I think.

Also this PR should be good to go. Here are the next steps:

- Merge the bundler lesson and this one (I'll handle this of course)
- Add the 2nd PR for the images in this lesson. You can follow our [guide for adding images](https://github.com/TheOdinProject/curriculum/wiki/Adding-Images-to-the-Curriculum) if you've never done this. Basically this adds the images to our CDN setup.
- PR to the main site repo that adds the lessons. You can find the guide for that on the [main site repo's wiki](https://github.com/TheOdinProject/theodinproject/wiki/Adding-a-Lesson)

After that, we should be done :partying_face:
